pkg/agent/controller/networkpolicy/reject.go

```diff
@@ -0,0 +1,250 @@
+	ICMPv6DstAdminProhibitedCode uint8 = 1
+)
+
+type RejectType int
+
+const (
+	// RejectPodLocal represents this packetOut is used to reject Pod-to-Pod traffic
+	// and for this response, the srcPod and the dstPod are on the same Node.
+	RejectPodLocal RejectType = iota
+	// RejectPodRemoteToLocal represents this packetOut is used to reject Pod-to-Pod
+	// traffic and for this response, the srcPod is on a remote Node and the dstPod is
+	// on the local Node.
+	RejectPodRemoteToLocal
+	// RejectLocalToRemote represents this packetOut is used to reject traffic and for
+	// this response, the srcPod is on the local Node and the dstPod is on a remote Node.
+	// While generating rejection from local to remote, there is no difference between
+	// Service traffic and Pod traffic.
+	RejectLocalToRemote
+	// RejectServiceLocal represents this packetOut is used to reject Service traffic,
+	// when AntreaProxy is enabled. The EndpointPod and the dstPod of the reject
+	// response are on the same Node.
+	RejectServiceLocal
+	// RejectServiceRemoteToLocal represents this packetOut is used to reject Service
+	// traffic, when AntreaProxy is enabled. The EndpointPod is on a remote Node and
+	// the dstPod of the reject response is on the local Node.
+	RejectServiceRemoteToLocal
+	// RejectNoAPServiceLocal represents this packetOut is used to reject Service
+	// traffic, when AntreaProxy is disabled. The EndpointPod and the dstPod of the
+	// reject response are on the same Node.
+	RejectNoAPServiceLocal
+	// RejectNoAPServiceRemoteToLocal represents this packetOut is used to reject
+	// Service traffic, when AntreaProxy is disabled. The EndpointPod is on a remote
+	// Node and the dstPod of the reject response is on the local Node.
+	RejectNoAPServiceRemoteToLocal
+)
+
+// rejectRequest sends reject response to the requesting client, based on the
+// packet-in message.
+func (c *Controller) rejectRequest(pktIn *ofctrl.PacketIn) error {
+	// Get ethernet data.
+	srcMAC := pktIn.Data.HWDst.String()
+	dstMAC := pktIn.Data.HWSrc.String()
+
+	var (
+		srcIP  string
+		dstIP  string
+		prot   uint8
+		isIPv6 bool
+	)
+	switch ipPkt := pktIn.Data.Data.(type) {
+	case *protocol.IPv4:
+		// Get IP data.
+		srcIP = ipPkt.NWDst.String()
+		dstIP = ipPkt.NWSrc.String()
+		prot = ipPkt.Protocol
+		isIPv6 = false
+	case *protocol.IPv6:
+		// Get IP data.
+		srcIP = ipPkt.NWDst.String()
+		dstIP = ipPkt.NWSrc.String()
+		prot = ipPkt.NextHeader
+		isIPv6 = true
+	}
+
+	sIface, srcFound := c.ifaceStore.GetInterfaceByIP(srcIP)
+	dIface, dstFound := c.ifaceStore.GetInterfaceByIP(dstIP)
+	// There are two situations in which it can be determined that this is a service
+	// traffic:
+	// 1. When AntreaProxy is enabled, EpSelectedRegMark is set in ServiceEPStateField.
+	// 2. When AntreaProxy is disabled, dstIP of reject response is on the local Node
+	//    and dstMAC of reject response is antrea-gw's MAC.
+	isServiceTraffic := func() bool {
```

Comment:
my understanding is that `isServiceTraffic` only return true if we are at the source Node (reject action in an ingress policy). Is that correct? For Service traffic captured at the destination Node (reject action in a egress policy), how is the traffic classified? 

Basically I don't understand how we can classify traffic a `RejectNoAPServiceRemoteToLocal` with this logic.

Reply:
I think I understand a bit better after reading the PR description again. But it still seems to me that in some cases (for egress policies), `isServiceTraffic` will return false for Service traffic, so it may be worth emphasizing this here with a comment, and explaining again that for such traffic the logic is the same as for Pod-to-Pod traffic (unless I am mistaken).

Reply:
Actually, it will only return true when the destination of the reject packet is on local Node. No matter in ingress/egress policy. 
I made a table for **AntreaProxy is disabled** hope it could help you understand or discover what I missed:

| Node       | Policy  | Traffic    | Rej Pkt                                                      | Type                           |
| ---------- | ------- | ---------- | ------------------------------------------------------------ | ------------------------------ |
| Inter-Node | Ingress | Pod-to-Pod | srcMAC: srcPod MAC<br />dstMAC: antrea-gw MAC<br />srcIP: srcPodIP<br />dstIP: dstPodIP(remote) | RejectLocalToRemote            |
| Inter-Node | Ingress | Service    | srcMAC: srcPod MAC<br />dstMAC: antrea-gw MAC<br />srcIP: srcPodIP<br />dstIP: dstPodIP(remote) | RejectLocalToRemote            |
| Inter-Node | Egress  | Pod-to-Pod | srcMAC: antrea-gw MAC<br />dstMAC: dstPod MAC<br />srcIP: srcPodIP<br />dstIP: dstPodIP(local) | RejectPodRemoteToLocal         |
| Inter-Node | Egress  | Service    | srcMAC: virtule MAC<br />dstMAC: antrea-gw MAC<br />srcIP: srcPodIP<br />dstIP: dstPodIP(local) | RejectNoAPServiceRemoteToLocal |
| Intra-Node | Ingress | Pod-to-Pod | srcMAC: srcPod MAC<br />dstMAC: dstPod MAC<br />srcIP: srcPodIP<br />dstIP: dstPodIP(local) | RejectPodLocal                 |
| Intra-Node | Ingress | Service    | srcMAC: srcPod MAC<br />dstMAC: antrea-gw MAC<br />srcIP: srcPodIP<br />dstIP: dstPodIP(local) | RejectNoAPServiceLocal         |
| Intra-Node | Egress  | Pod-to-Pod | srcMAC: srcPod MAC<br />dstMAC: dstPod MAC<br />srcIP: srcPodIP<br />dstIP: dstPodIP(local) | RejectPodLocal                 |
| Intra-Node | Egress  | Service    | srcMAC: srcPod MAC<br />dstMAC: antrea-gw MAC<br />srcIP: srcPodIP<br />dstIP: dstPodIP(local) | RejectNoAPServiceLocal         |

Reply:
> I think I understand a bit better after reading the PR description again. But it still seems to me that in some cases (for egress policies), isServiceTraffic will return false for Service traffic, so it may be worth emphasizing this here with a comment, and explaining again that for such traffic the logic is the same as for Pod-to-Pod traffic (unless I am mistaken).

Do you mean inter-Node ingress cases? Pod-to-Pod traffic and Service traffic have the same logic?

Reply:
I added more comments here to explain this. @antoninbas you can take a look to see if it's clear. Thanks.

Reply:
In my original comment, I switched "ingress" and "egress" so it must have been confusing :/

> Actually, it will only return true when the destination of the reject packet is on local Node. No matter in ingress/egress policy.

I see that, but it is always the case for egress policies, and often not the case for ingress policies (inter-Node). Thanks for the table. I think you should copy the source code for the table to the PR description for future reference.

Reply:
Good idea. Added to the PR description.